lib/legacy/zstd_v03.c

```diff
@@ -2711,7 +2711,7 @@ static size_t ZSTD_execSequence(BYTE* op,
     if (litEnd > litLimit) return ERROR(corruption_detected);   /* overRead beyond lit buffer */
 
     /* copy Literals */
-    ZSTD_wildcopy(op, *litPtr, sequence.litLength);   /* note : oLitEnd <= oend-8 : no risk of overwrite beyond oend */
+    ZSTD_memmove(op, *litPtr, sequence.litLength);   /* note : used to be wildCopy, changed to fix a bug in 32-bit mode (oss-fuzz case 55714) */
```

Comment:
I don't think this fix is sufficient. We're supposed to have 8 extra bytes in the litBuffer, so this should be valid as long as `litEnd <= litLimit`.

I think the problem we're running into is that the limit checks should be re-written to avoid overflow. We don't care about speed, so we should do the obviously correct thing.

The offset check also looks janky, so we should clean that up as well.

E.g.

```c
size_t const seqLength = sequence.litLength + sequence.matchLength;

if (seqLength > (size_t)(oend - op)) return ERROR(dstSize_tooSmall);
if (sequence.litLength > (size_t)(litLimit -*litPtr)) return ERROR(corruption_detected);
/* Now we know we don't have any overflow in literal / match lengths, can use the pointer check for oend_8*/
if (oLitEnd > oend_8) return ERROR(dstSize_tooSmall);
if (sequence.offset > (U32)(oLitEnd - base)) return ERROR(corruption_detected);
```

This bug is present in every version of the legacy decoder, so we'll have to fix every one.